### PR TITLE
fix(deps): [sc-32782] Fix action to show correct metrics

### PR DESCRIPTION
**Pull request by donleistman**
Shortcut: https://app.shortcut.com/summer/story/32782

## What

Fix issues with our shared action to send metrics on % of out-of-date dependencies to datadog.

## Why

Couple of issues here:
- One, this was written with the assumption that `yarn outdated` was listing all dependencies in a project (as per the docs), however the command only lists outdated dependencies. So the total # of deps we were using was wrong.
- We were parsing the semver incorrectly, which also threw everything off.

## How

- Add an additional command to parse the package.json and build the total number of dependencies ourselves
- Add `.split('.')` to the semver string so we can destructure it correctly, and convert strings to numbers so we can compare correctly

Also increased the precision of the percentages -- this should ensure that they add up to 100%, or at least get close enough that we don't see a wavy line at the top of the graph.

This will only work in single-package repos. I made a PR in the monorepo to stop using this action: https://github.com/simplifidev/summer/pull/2078. We'll need a separate solution for the monorepo.

## How to Test

Here's the before and after of the output for each repo:

### shared

**before:**
Screen Shot 2022-05-03 at 7 10 04 PM

**after:**
Screen Shot 2022-05-03 at 7 09 44 PM

### server

**before:**
Screen Shot 2022-05-03 at 7 24 18 PM

**after:**
Screen Shot 2022-05-03 at 7 37 08 PM

### devbot

**before:**
Screen Shot 2022-05-03 at 7 39 04 PM

**after:**

Screen Shot 2022-05-03 at 7 38 13 PM

### CODE REVIEW

File: calcOutdatedMetrics.js
Comment (JonathanVanDyke) [inline calcOutdatedMetrics.js:1]: huh, so this is an internal node package that's available after the github action setup?
Is there a way to pin the node version in the action yml?

File: calcOutdatedMetrics.js
Comment (JonathanVanDyke) [inline calcOutdatedMetrics.js]: Possible to refactor into methods?
Maybe something like `function dataSetup` etc, then have just one main fn kick off all the logic?

File: calcOutdatedMetrics.js
Comment (JonathanVanDyke) [inline calcOutdatedMetrics.js]: Love the parallel assignment,
though I got burned recently doing something similar with the output of sequelize update since the size of the return array changes based on what's updated

Do these arrays here guarantee a return size/structure/ordering?

File: calcOutdatedMetrics.js
Comment (JonathanVanDyke) [inline calcOutdatedMetrics.js]: might be able to flatten the deep conditional nest here, nit tho

fn flattening
``` javascript
const updateVersions = (behindMajor, behindMinor, currentPatch, package) => {
  if (currentMajor < latestMajor) {
    behindMajor.push(package)
    return
  }
  
  if (currentMinor < latestMinor ) {
    behindMinor.push(package)
    return
  }
  
  if (currentPatch < latestPatch) {
    behindPatch.push(package)
    return
  }
  
  throw new Error('package ...');
}
```

dirty inline flattening
```javascript
const major = [currentMajor, latestMajor, behindMajor]
const minor = [currentMinor, latestMinor, behindMinor]
const patch = [currentPatch, latestPatch, behindPatch]

const version = [major, minor, patch]
for (let i = 0; i < version.length; i++) {
  const [current, latest, behind] = version[0]
  if (current < latest) {
    behind.push(package)
    break
  }

  i == version.length - 1 ? throw new Error('package ....') : null
}
```

File: calcOutdatedMetrics.js
Comment (JonathanVanDyke) [inline calcOutdatedMetrics.js:56]: overkill but interesting solution in t[his article](https://www.delftstack.com/howto/javascript/javascript-round-to-2-decimal-places/#:~:text=instead%20of%202.01%20.-,Use%20the%20Math.,ensure%20the%20number's%20accurate%20rounding.) for the rounding problem

their solution:
```javascript
function roundToTwo(num) {
    return +(Math.round(num + "e+2")  + "e-2");
}
console.log(roundToTwo(2.005));
```

ofc your solution is good, just adding for curiosities sake

File: calcOutdatedMetrics.js
Comment (calvin-cdev) [reply calcOutdatedMetrics.js:1]: if it's internal then it's controlled by the node version

**Review by calvin-cdev — COMMENTED** [review]
going to leave this to @JonathanVanDyke 
i'll note that if we're concerned about weird JS float issues, you could pull in the package `decimal.js` and use it for the math.

File: calcOutdatedMetrics.js
Comment (JonathanVanDyke) [reply calcOutdatedMetrics.js:1]: where's the node version defined though?

Comment (calvin-cdev) [reply calcOutdatedMetrics.js:1]: good question 🤔

Comment (JonathanVanDyke) [reply calcOutdatedMetrics.js:1]: *updates to `node 19`, bricks all the code
lol

Comment (donleistman) [reply calcOutdatedMetrics.js:1]: Found [this](https://github.blog/changelog/2021-12-10-github-actions-github-hosted-runners-now-run-node-js-16-by-default/) and [this](https://github.com/actions/setup-node#usage) -- I'll try pinning it with the node setup action.

As a side note, this whole thing is kind of a frankenstein -- I've got `yarn outdated` piping into this node script, then the node script is spinning up a separate bash script to read the package.json. And then that's all getting piped to another bash script to make the curl request to datadog.

The better way to do it would probably to have the action execute the `yarn` and package.json commands separately, store the outputs, then pass them into the node script as parameters. But the current way should work and I don't want to refactor it  too heavily just yet. 

I'm thinking I'll try another approach with the monorepo action, because that's going to need a bunch more inputs (all the package.json), and if it works there maybe I'll come back and refactor this one.

File: calcOutdatedMetrics.js
Comment (donleistman) [reply calcOutdatedMetrics.js]: This is the output from `yarn outdated --json`. Try running it yourself, with and without the json flag, and that should give you a good idea of what we're looking at.

Here's the output from the shared repo:
```json
{
  "type": "table",
  "data": {
    "head": ["Package", "Current", "Wanted", "Latest", "Package Type", "URL"],
    "body": [
      [
        "@types/jest",
        "27.4.1",
        "27.4.1",
        "27.5.0",
        "devDependencies",
        "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jest"
      ],
      [
        "@typescript-eslint/eslint-plugin",
        "5.21.0",
        "5.21.0",
        "5.22.0",
        "devDependencies",
        "https://github.com/typescript-eslint/typescript-eslint#readme"
      ],
      [
        "@typescript-eslint/parser",
        "5.21.0",
        "5.21.0",
        "5.22.0",
        "devDependencies",
        "https://github.com/typescript-eslint/typescript-eslint#readme"
      ],
      [
        "faker",
        "5.5.3",
        "5.5.3",
        "6.6.6",
        "dependencies",
        "https://github.com/Marak/Faker.js#readme"
      ],
      [
        "jest",
        "27.5.1",
        "27.5.1",
        "28.0.3",
        "devDependencies",
        "https://jestjs.io/"
      ],
      [
        "jest-mock",
        "27.5.1",
        "27.5.1",
        "28.0.2",
        "dependencies",
        "https://github.com/facebook/jest#readme"
      ],
      [
        "plaid",
        "10.1.0",
        "10.1.0",
        "10.3.0",
        "dependencies",
        "https://github.com/plaid/plaid-node#readme"
      ],
      [
        "ts-jest",
        "27.1.4",
        "27.1.4",
        "28.0.1",
        "devDependencies",
        "https://kulshekhar.github.io/ts-jest"
      ]
    ]
  }
}
```

It looks like `--json` flag does a something of a crude serialization of the table from the default output, which is why it has these ordered arrays instead of something more intuitive like an object for each dependency.

File: calcOutdatedMetrics.js
Comment (donleistman) [reply calcOutdatedMetrics.js]: Nice I will try it out.

**Comment by donleistman** [comment]
> going to leave this to @JonathanVanDyke i'll note that if we're concerned about weird JS float issues, you could pull in the package `decimal.js` and use it for the math.

I'm not really concerned about them because whether the numbers add up to 100% or 100.1% won't affect our ability to parse the graph at all.

File: calcOutdatedMetrics.js
Comment (donleistman) [reply calcOutdatedMetrics.js]: Hmmm can you maybe describe more what this may look like?

The way I'm constructing `data` here is due to the fact that the `yarn outdated` output is being piped in, so it's a stream. If we refactor later to pass it in as an argument instead (as per other comment) we may be able to get rid of this.

File: action.yml
Comment (JonathanVanDyke) [inline action.yml:13]: 👍🏼

File: calcOutdatedMetrics.js
Comment (JonathanVanDyke) [reply calcOutdatedMetrics.js]: sure, kinda just refactoring into bite sized fns

Here's the gist of it (lol pun): https://gist.github.com/JonathanVanDyke/af9bca709b17f14d998a2126eb86d454
idk if you can open it tho since it's secret?
this all might be overkill if this is just in the early prototype stage tho, also got lost in the sauce a bit so it's prob not compilable code - just meant for reference

basically it kicks itself off with a self invoked fn up top to kind of hide the global scope a bit, in python I know they use something like `if __name__ === filename {do the things}` when it's a more simple single file script, so reaching for an equivalent here

File: calcOutdatedMetrics.js
Comment (JonathanVanDyke) [reply calcOutdatedMetrics.js]: oh sweet, i'll try it!

File: calcOutdatedMetrics.js
Comment (donleistman) [reply calcOutdatedMetrics.js]: I think we still need `data` to be in the global scope. But I refactored a bit see what you think.

**Review by JonathanVanDyke — APPROVED** [review]
Beautiful! 🕺🏼 🎊

**Comment by JonathanVanDyke** [comment]
ah I don't have the power to approve on this repo maybe? 🤔

**Comment by donleistman** [comment]
hmmm I'll take a look

**Comment by donleistman** [comment]
@JonathanVanDyke I had myself as the only one with write access on the repo, just added the rest of the eng team.

**Pull request merged**
